### ISSUE TITLE
Vulnerable Regular Expression

### ISSUE THREAD

**Issue by cristianstaicu** [lodash/lodash#3359]
The following regular expression used in words.js is vulnerable to ReDoS:
```js
/[a-z][A-Z]\|[A-Z]{2,}[a-z]\|[0-9][a-zA-Z]\|[a-zA-Z][0-9]\|[^a-zA-Z0-9 ]/
```
This regex is called by multiple methods exposed to the users such as "camelCase". The slowdown is moderately low: for 50.000 characters around 2 seconds matching time. However, I would still suggest one of the following: 
- remove the regex,
- anchor the regex,
- limit the number of characters that can be matched by the repetition,
- limit the input size.

If needed, I can provide an actual example showing the slowdown.

**Comment by cristianstaicu** [comment]
I agree that for the general case that is a very large value, but security is about corner cases. Let us consider that for the camel case export alone (https://www.npmjs.com/package/lodash.camelcase) there are 237 dependents. I am pretty sure it can produce unexpected results in some packages. For example this package which seems to be the most popular one:
https://www.npmjs.com/package/css-loader

It looks like it just applies the camel case package on random CSS files you can specify using URLs. I am pretty sure there are nastier things than this, such as parsing HTTP headers which an attacker can easily control. 

I would like to submit a PR, but you have to tell me if any of the following options are acceptable: 
1) transform the regex into :
```js
/[a-z][A-Z]\|[A-Z]{2,2000}[a-z]\|[0-9][a-zA-Z]\|[a-zA-Z][0-9]\|[^a-zA-Z0-9 ]/
```
This effectively limits the word size to 2000 characters

2) check that the string parameter in "words" function has at most 20,000 characters. 

If, as you say, most of the use cases involve fewer number of characters, this should impact none of these use cases, but my prediction is that soon you will get new bug reports from angry users that random errors pop up in their application. :) Any other ideas?

**Comment by jdalton** [comment]
Limiting the size is fine if there's no other way. The regexp is a monster so if there's a way to improve that, I'd prefer that route.

**Comment by gibson042** [comment]
> The regexp is a monster so if there's a way to improve that, I'd prefer that route.

Well, it seems to break down into the following alternatives:
* lowercase a-z, followed by uppercase
* two or more uppercase A-Z, followed by lowercase
* 0-9 followed by A-Z (any case)
* A-Z (any case) followed by 0-9
* any character that is not a space or 0-9 or A-Z (any case)

So the most straightforward simplification just drops the unbounded quantifier and introduces more negated classes:
```js
// only tweak number-adjacent classes:
/[a-z][A-Z]|[A-Z]{2}[a-z]|[0-9][^0-9 ]|[^0-9 ][0-9]|[^a-zA-Z0-9 ]/

// or tweak all three prefixes:
/[a-z][^a-z ]|[A-Z][0-9]|[A-Z]{2}[^A-Z ]|[0-9][^0-9 ]|[^a-zA-Z0-9 ]/
```
It's still ugly, but that seems hard to avoid given the lowercase-after-one-uppercase exception.

**Comment by viralanomaly** [comment]
Is there a plan to limit the size?  The file in question has not changed since well before this issue was raised.

**Comment by manuel-jasso** [comment]
Hi @jdalton, my company is tightening security and this issue is blocking us at the moment. If checking the string parameter length is an acceptable solution, would you accept a PR with this change?

**Comment by jdalton** [comment]
I'm still inclined to think of this issues as more or less expected. By that I mean if you're running `camelCase` over 50kB of text it's reasonable that it's not going to be great perf. Perf tends to decrease with larger payloads.

> If checking the string parameter length is an acceptable solution, would you accept a PR with this change?

I'd prefer improving the performance of the existing regexp to an arbitrary limit. The limit would be if no other solution can be made.

**Comment by manuel-jasso** [comment]
Thanks for responding 😄 .

> I'm still inclined to think of this issues as more or less expected.

Totally agree, however the security team is asking me "can you guarantee that nobody will pass large strings" and I can guarantee that for my own code, but I cannot guarantee that for third party code unless the library itself does that.

I know this would be a larger change, but how about adding some kind of opt-in config parameter to loadsh so that it works in a more secure mode at run time, even though it would be a little slower? That would be a good trade-off between security and performance.

If you prefer to fix the RegEx, are the changes proposed by @gibson042 acceptable?

**Comment by jdalton** [comment]
> Totally agree, however the security team is asking me "can you guarantee that nobody will pass large strings" and I can guarantee that for my own code [...] how about adding some kind of opt-in config parameter to loadsh so that it works in a more secure mode at run time, even though it would be a little slower? 

You're probably better off using your own methods for this.

> If you prefer to fix the RegEx, are the changes proposed by @cristianstaicu or @gibson042 acceptable?

 No idea. You'll want to patch against [4.17.10](https://github.com/lodash/lodash/tree/4.17.10) and ensure unit tests still pass.

**Comment by ccanning** [comment]
The regex only performs poorly when you get a large number of consecutive capital letters. Using regex 101, I can timeout your regex with 50K A's. if i switch the regex to non-greedy, it at least completes, but still takes 1.3-1.6 seconds.

`  /[a-z][A-Z]|[A-Z][A-Z]+?[a-z]|[0-9][a-zA-Z]|[a-zA-Z][0-9]|[^a-zA-Z0-9 ]/;`

if you remove the `[A-Z][A-Z]+?[a-z]` matching section, its back to 1ms matching. Since this is used in one place in your code and matches a special case, you could possibly do 2 regex's and only test the second if the first fails. 

If this is not an option, then the slight change does make your regex more resilient.

**Comment by ccanning** [comment]
Looks like i broke the online tool, and it is incorrectly matching the consecutive series of capital letters. So, you can disregard those comments, but the change in regex should be more resilient.

**Comment by ccanning** [comment]
I think I finally have a fully-working regex that guards against the all caps attack and still uses your original matching patterns, I just changed the method to **positive lookahead** with **non-greedy** matching. It also doesn't have an arbitrary string length limit. 

``` js 
/[a-z][A-Z]|(?<=[A-Z][A-Z]+?[a-z])|[0-9][a-zA-Z]|[a-zA-Z][0-9]|[^a-zA-Z0-9 ]/; 
```

**Comment by thinkkevin** [comment]
‘[A-Z]{2,}[a-z]’ makes sense when using Regex to match a string greedily. BUT it is a waste if you just want to use such regex to test a string. It is same as ‘[A-Z]{2}[a-z]’ for the regex.test purpose.

**Comment by manuel-jasso** [comment]
Hi @jdalton, as you can see there is quite some excitement and interest to solve this issue! 😄 

This is work in progress using the simplest RegEx with the smallest possible change (removing the comma):
[manuel-jasso:prevent-redos#diff-001d0647fb00f8336795faccdec19a31R280](https://github.com/manuel-jasso/lodash/compare/4.17.10...manuel-jasso:prevent-redos#diff-001d0647fb00f8336795faccdec19a31R280)

I added some tests ro prove `_.words` can still parse huge words successfully, especially the kind of words that use the slow performing RegEx alternative, and it can do it in less than 250ms (I'm testing two cases and asserting it takes less than 500ms):
[manuel-jasso:prevent-redos#diff-c1129c8b045390789fa8ff62f2c6b4a9R25211](https://github.com/manuel-jasso/lodash/compare/4.17.10...manuel-jasso:prevent-redos#diff-c1129c8b045390789fa8ff62f2c6b4a9R25211)

What do you think so far?

**Comment by DyzLecticus** [comment]
Has the fix for this issue been merged and released?

**Comment by jdalton** [comment]
Hi @DyzLecticus!

Yes. It's patched in v4.17.11.

**Comment by ddillard** [comment]
FYI, I submitted to get a CVE for this issue.  I'll add a comment once the CVE ID has been issued.
